examples/async-critical/src/wasm.rs

```diff
@@ -0,0 +1,112 @@
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::HandleAction;
+use gstd::{critical, exec, msg, prelude::*};
+
+static mut REPLY_SET_HOOK: bool = false;
+static mut SIGNAL_SET_HOOK: bool = false;
+
+#[gstd::async_main(handle_reply = my_handle_reply, handle_signal = my_handle_signal)]
+async fn main() {
+    let action: HandleAction = msg::load().expect("Failed to read handle action");
+
+    match action {
+        HandleAction::Simple => {
+            // call `gr_source` outside because it is forbidden in `handle_signal`
+            let source = msg::source();
+
+            // should not send anything because execution will be completed
+            critical::set_hook(move || {
+                msg::send_bytes(source, b"critical", 0).unwrap();
+            });
+
+            // wait occurs inside so hook is saved
+            gstd::msg::send_bytes_for_reply(source, b"for_reply", 0, 0)
+                .expect("Failed to send message")
+                .await
+                .expect("Received error reply");
+        }
+        HandleAction::Panic => {
+            // call `gr_source` outside because it is forbidden in `handle_signal`
+            let source = msg::source();
+
+            // should send message because panic occurs below
+            critical::set_hook(move || {
+                msg::send_bytes(source, b"critical", 0).unwrap();
+            });
+
+            // wait occurs inside so hook is saved
+            gstd::msg::send_bytes_for_reply(msg::source(), b"for_reply", 0, 0)
+                .expect("Failed to send message")
+                .await
+                .expect("Received error reply");
+
+            // panic occurs so `handle_signal` will execute hook
+            panic!();
+        }
+        HandleAction::InHandleReply => {
+            unsafe {
+                REPLY_SET_HOOK = true;
+            }
+
+            gstd::msg::send_bytes_for_reply(msg::source(), b"for_reply", 0, 0)
+                .expect("Failed to send message")
+                .await
+                .expect("Received error reply");
+        }
+        HandleAction::InHandleSignal => {
+            unsafe {
+                SIGNAL_SET_HOOK = true;
+            }
+
+            gstd::msg::send_bytes_for_reply(msg::source(), b"for_reply", 0, 0)
+                .expect("Failed to send message")
+                .await
+                .expect("Received error reply");
+
+            panic!()
+        }
+    }
+}
+
+fn my_handle_reply() {
+    unsafe {
+        if REPLY_SET_HOOK {
+            let source = msg::source();
+
+            // should panic in this entrypoint
+            critical::set_hook(move || {
+                msg::send_bytes(source, b"from_handle_reply", 0).unwrap();
+            });
+        }
+    }
+}
+
+fn my_handle_signal() {
+    unsafe {
+        if SIGNAL_SET_HOOK {
+            let source = msg::source();
```

Comment:
aren't we forbid taking source for handle signal?

Reply:
Nice catch